src/mpl/include/mpl_atomic_opa.h

```diff
@@ -0,0 +1,218 @@
+{                                                                              \
+    return (TYPE)OPA_load_acquire_int(&ptr->v);                                \
+}                                                                              \
+static inline void MPL_atomic_relaxed_store_ ## NAME                           \
+                            (struct MPL_atomic_ ## NAME ## _t * ptr, TYPE val) \
+{                                                                              \
+    OPA_store_int(&ptr->v, (int)val);                                          \
+}                                                                              \
+static inline void MPL_atomic_release_store_ ## NAME                           \
+                            (struct MPL_atomic_ ## NAME ## _t * ptr, TYPE val) \
+{                                                                              \
+    OPA_store_release_int(&ptr->v, (int)val);                                  \
+}                                                                              \
+static inline TYPE MPL_atomic_cas_ ## NAME                                     \
+                (struct MPL_atomic_ ## NAME ## _t * ptr, TYPE oldv, TYPE newv) \
+{                                                                              \
+    return (TYPE)OPA_cas_int(&ptr->v, (int)oldv, (int)newv);                   \
+}                                                                              \
+static inline TYPE MPL_atomic_swap_ ## NAME                                    \
+                            (struct MPL_atomic_ ## NAME ## _t * ptr, TYPE val) \
+{                                                                              \
+    return (TYPE)OPA_swap_int(&ptr->v, (int)val);                              \
+}                                                                              \
+static inline TYPE MPL_atomic_fetch_add_ ## NAME                               \
+                            (struct MPL_atomic_ ## NAME ## _t * ptr, TYPE val) \
+{                                                                              \
+    return (TYPE)OPA_fetch_and_add_int(&ptr->v, (int)val);                     \
+}                                                                              \
+static inline TYPE MPL_atomic_fetch_sub_ ## NAME                               \
+                            (struct MPL_atomic_ ## NAME ## _t * ptr, TYPE val) \
+{                                                                              \
+    return (TYPE)OPA_fetch_and_add_int(&ptr->v, -(int)val);                    \
+}
+
+/* int */
+MPL_ATOMIC_DECL_FUNC_INT32(int, int)
+/* int32_t */
+MPL_ATOMIC_DECL_FUNC_INT32(int32_t, int32)
+/* uint32_t */
+MPL_ATOMIC_DECL_FUNC_INT32(uint32_t, uint32)
+
+/* int64_t and uint64_t */
+struct MPL_atomic_int64_static_assert_t {
+    /* The following causes an error if the assumption does not hold. */
+    int assert1[sizeof(void *) == 8 ? 1 : -1];
```

Comment:
Similarly, what happens when `sizeof(void *)` is 4?

Reply:
An compilation error happens.

Reply:
So what is your solution on 32-bit system?

Reply:
For now there's no way to support it (OPA doesn't support it). We should fallback to lock-based atomics, but that is user's responsibility since this OPA wrapper is available only when a user sets it explicitly.

Reply:
Users set it explicitly because they know exactly what openpa do -- supports int. You are surprising user with no back-up plans.

Reply:
What is your suggestion? Do we need to detect it at configuration time instead of compile time?

Reply:
I was suggesting make it a compile-time error for sized-int atomics in the case of explicit openpa primitives.

Reply:
https://github.com/pmodels/mpich/blob/master/src/mpid/ch4/shm/posix/release_gather/release_gather.h#L382 uses sized-int atomics.

I'm fine to make it compile-time error, but it really narrows applicability. Is it okay for PGI support?

Reply:
> https://github.com/pmodels/mpich/blob/master/src/mpid/ch4/shm/posix/release_gather/release_gather.h#L382 uses sized-int atomics.

I have to read more to understand that code's purpose. But if that is supposed to be an address, then they really should use ptr type rather than `uint64_t`, right?

Reply:
Yeah, so I re-enforced my belief to make it a compile-time error, so we can get an opportunity/reason to push-back on developers and ask them to really think what they are trying to do :)

Reply:
I can find pretty many `atomic_uint64_t` in `ch4`. Discouraging use of sized-int atomics to support OPA (which is fundamentally for PGI compilers (which becomes problematic only on a 32-bit machine)) does not sound right.

In addition, developers should not be aware of the underlying atomic implementation and that is why an atomic wrapper was introduced. I believe its functionality and semantics should be as uniform as possible regardless of whether developers (actually users) explicitly choose `openpa` or not.

---

However, I understand your concern; this PR is not doing the best. It overlooks many possible cases. Current:
- It supports only `sizeof(int) == 4` and `sizeof(void *) == 8`.

However, as you mentioned, we can support the following too.
- `sizeof(int) >= 8` or `sizeof(void *) >= 8`.

For example, the current wrapper supports neither `sizeof(int) == 1 && sizeof(void *) == 8` nor `sizeof(int) == 8 && sizeof(void *) == 4`. If we should support such cases (though I cannot come up with such a system), please let me know.

I note that there is no way to implement this MPL/atomic wrapper if `sizeof(int) < 8` and `sizeof(void *) < 8`.  Should we automatically fall back to the emulation lock in that case?